### PR TITLE
Balder/replace homegrown atomics with std atomic

### PR DESCRIPTION
**Pull request by baldersheim**
@vekterli PR
@toregge FYI

### CODE REVIEW

File: fnet/src/vespa/fnet/frt/rpcrequest.cpp [hunk 1]
```diff
@@ -129,7 +129,7 @@ void
 FRT_RPCRequest::SubRef()
 {
     assert(_refcnt > 0);
-    if (vespalib::Atomic::postDec(&_refcnt) == 1) {
+    if (_refcnt.fetch_sub(1) == 1) {
```
Comment (vekterli) [inline fnet/src/vespa/fnet/frt/rpcrequest.cpp]: Both the implicit atomic load in the `assert` and the `fetch_sub` will trigger full fences. If this is a problem, can assert on the returned value from `fetch_sub` instead to reduce fence count by 1.

**Review by vekterli — APPROVED** [review]
LGTM, with a few style consistency comments.

With default `fetch_add`/`fetch_sub` memory ordering, the `std::atomics` code may have stronger memory barriers in place than what was the case with the old FastOS atomics code. In particular, all such calls will incur a full compiler and CPU memory load/store fence. This is not really an issue unless we notice that it's causing performance regressions (after all, fences generally make threaded behavior easier to reason about).

If we do want to change memory ordering, counter updates and `ReferenceCounter` are low-hanging fruit.

**Comment by baldersheim** [comment]
I have fixed your comments apart from the optimisations stuff.

File: fnet/src/vespa/fnet/frt/rpcrequest.cpp [hunk 2]
```diff
@@ -128,8 +128,9 @@ FRT_RPCRequest::Recycle()
 void
 FRT_RPCRequest::SubRef()
 {
-    assert(_refcnt > 0);
-    if (vespalib::Atomic::postDec(&_refcnt) == 1) {
+    int oldVal = _refcnt.fetch_sub(1);
+    assert(oldVal > 1);
```
Comment (toregge) [inline fnet/src/vespa/fnet/frt/rpcrequest.cpp]: Should be > 0 ?

**Pull request merged**